### PR TITLE
Modified for flatlist (from #629)

### PR DESCRIPTION
**Pull request by xcarpentier**
From https://github.com/FaridSafi/react-native-gifted-chat/pull/629

### CODE REVIEW

**Comment by xcarpentier** [comment]
I think we need to have an expo publish on every PR to test easely...

**Comment by cooperka** [comment]
Thanks for cleaning this up @xcarpentier :beers: But fyi I probably won't have time to review this week, you may want to find someone else to review.

**Comment by brunocascio** [comment]
Great work @xcarpentier !

I am in vacation now, but I saw part of the code. Is there a way to remove internal state and use only props un order to make this lib completely stateless. What do you think?

File: src/MessageContainer.js [hunk 1]
```diff
@@ -95,72 +86,66 @@ export default class MessageContainer extends React.Component {
...
-
-  renderScrollComponent(props) {
-    const { invertibleScrollViewProps } = this.props;
     return (
-      <InvertibleScrollView
-        {...props}
-        {...invertibleScrollViewProps}
-        ref={(component) => (this._invertibleScrollViewRef = component)}
-      />
+      <View style={{ transform: [{ scaleY: -1 }, { perspective: 1280 }] }}>
```
Comment (brunolemos) [inline src/MessageContainer.js]: Why this?

File: src/MessageContainer.js [hunk 2]
```diff
@@ -95,72 +86,66 @@ export default class MessageContainer extends React.Component {
...
+  render() {
     return (
       <View ref="container" style={styles.container}>
-        <ListView
+        <FlatList
           enableEmptySections
           automaticallyAdjustContentInsets={false}
           initialListSize={20}
           pageSize={20}
+          ref="flatListRef"
```
Comment (brunolemos) [inline src/MessageContainer.js]: String ref was deprecated a year ago, you should use callback refs only.

```jsx
ref={flatListRef =>
  if (flatListRef) {
    this.flatListRef = flatListRef
  }
}
```

File: src/MessageContainer.js [hunk 3]
```diff
@@ -95,72 +86,66 @@ export default class MessageContainer extends React.Component {
...
-          contentContainerStyle={contentContainerStyle}
-          renderRow={this.renderRow}
-          renderHeader={this.props.inverted ? this.renderFooter : this.renderLoadEarlier}
-          renderFooter={this.props.inverted ? this.renderLoadEarlier : this.renderFooter}
-          renderScrollComponent={this.renderScrollComponent}
+          data={this.state.messagesData}
+          renderItem={this.renderRow}
+          renderHeader={this.renderFooter}
+          renderFooter={this.renderLoadEarlier()}
+          style={{ transform: [{ scaleY: -1 }, { perspective: 1280 }] }}
```
Comment (brunolemos) [inline src/MessageContainer.js]: Why this?

File: src/MessageContainer.js [hunk 4]
```diff
@@ -172,8 +157,9 @@ const styles = StyleSheet.create({
   container: {
     flex: 1,
   },
-  notInvertedContentContainerStyle: {
-    justifyContent: 'flex-end',
+  headerWrapper: {
+    flex: 1,
+    transform: [{ scaleY: -1 }, { perspective: 1280 }],
```
Comment (brunolemos) [inline src/MessageContainer.js]: Why this?

File: src/__tests__/__snapshots__/MessageContainer.test.js.snap [hunk 5]
```diff
@@ -9,21 +9,70 @@ exports[`should render <MessageContainer /> and compare with snapshot 1`] = `
...
+    onContentSizeChange={[Function]}
+    onEndReachedThreshold={2}
+    onLayout={[Function]}
+    onMomentumScrollEnd={[Function]}
+    onScroll={[Function]}
+    onScrollBeginDrag={[Function]}
+    onScrollEndDrag={[Function]}
     pageSize={20}
-    renderFooter={[Function]}
+    renderFooter={null}
```
Comment (brunolemos) [inline src/__tests__/__snapshots__/MessageContainer.test.js.snap]: Why this?

File: src/__tests__/__snapshots__/MessageContainer.test.js.snap [hunk 6]
```diff
@@ -9,21 +9,70 @@ exports[`should render <MessageContainer /> and compare with snapshot 1`] = `
...
+    scrollEventThrottle={50}
+    stickyHeaderIndices={Array []}
+    style={
+      Object {
+        "transform": Array [
+          Object {
+            "scaleY": -1,
+          },
+          Object {
+            "perspective": 1280,
```
Comment (brunolemos) [inline src/__tests__/__snapshots__/MessageContainer.test.js.snap]: Why this?

File: src/__tests__/__snapshots__/MessageContainer.test.js.snap [hunk 7]
```diff
@@ -9,21 +9,70 @@ exports[`should render <MessageContainer /> and compare with snapshot 1`] = `
...
+        <View
+          style={
+            Object {
+              "flex": 1,
+              "transform": Array [
+                Object {
+                  "scaleY": -1,
+                },
+                Object {
+                  "perspective": 1280,
```
Comment (brunolemos) [inline src/__tests__/__snapshots__/MessageContainer.test.js.snap]: Why this?

**Comment by brunolemos** [comment]
I found very weird that the code `transform: [{ scaleY: -1 }, { perspective: 1280 }]` appears 6 times in this PR and I'm not sure why they are necessary. Maybe this can be optimized? Implemented differently?

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
Thanks, @brunolemos I removed this style `transform: [{ scaleY: -1 }, { perspective: 1280 }]`

**Comment by xcarpentier** [comment]
Thanks @brunocascio, I removed state and extends PureComponent

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by brunolemos** [comment]
This last one is working fine.

When there are dozens of messages and I scroll to the top fast, it takes some seconds to load the previous items. I was expecting it to be instant as I don't think it's using any server on this example. Or is it?

**Comment by xcarpentier** [comment]
No server.
But I need to figured out what is the best configuration for this FlatList.

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
@brunolemos I changed props on FlatList, please try and let me know
```javascript
<FlatList
  initialNumToRender={10}
  maxToRenderPerBatch={10}
  windowSize={10}
/>
```

**Comment by brunolemos** [comment]
@xcarpentier Check this:

gif

Maybe it's not related to this PR?

**Comment by xcarpentier** [comment]
@brunolemos I think it is, thanks you.
Will dig more into it...

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

File: src/MessageContainer.js [hunk 8]
```diff
@@ -107,7 +107,7 @@ export default class MessageContainer extends React.PureComponent {
           initialNumToRender={MAX_RENDER_ITEM}
           maxToRenderPerBatch={MAX_RENDER_ITEM}
           windowSize={MAX_RENDER_ITEM}
-          removeClippedSubviews
+          removeClippedSubviews={false}
```
Comment (brunolemos) [inline src/MessageContainer.js]: Did you face any problem? Using false has performance implications.

Comment (xcarpentier) [reply src/MessageContainer.js]: The doc mention bug related to this props so I try to see that change but no https://facebook.github.io/react-native/docs/flatlist.html#removeclippedsubviews

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by Shelf-A** [comment]
Awesome work @xcarpentier.

Thanks for modifying for flatlist. The last PR looks great but everytime the list gets larger the keyboard freezes and the chat input gets slower too and then the chats disappear momentarily when you scroll up . I don't know if this is caused by Expo...

**Comment by xcarpentier** [comment]
@Shelf-A thank you very much for your feedbacks.
iOS or Android?

**Comment by Shelf-A** [comment]
Android

**Comment by brunocascio** [comment]
@xcarpentier do you need some help here?

**Comment by xcarpentier** [comment]
@brunocascio YES, I don't find time to figured out what's going on with "android" issues when large list!
We need to find best FlatList props definition I guess?

**Comment by brunocascio** [comment]
@xcarpentier OK. I will try to figure out what's going on later 👍

**Comment by Shelf-A** [comment]
@xcarpentier @brunocascio I can pinpoint few issues.

1. Constant.js sets a static MAX_RENDER_ITEM to 25 which is used for initialNumToRender and according to https://facebook.github.io/react-native/docs/flatlist.html#initialnumtorender we don't need to render 25 items. maxToRenderPerBatch is MAX_RENDER_ITEM * 2 which gives us 50 which is good for fill rate but performance not so good cos the send button freezes, same goes for windowSize

2. InputToolbar.js this component re-renders many times. It re-renders 7 times when the keyboard is shown and 7 more times when the keyboard is dismissed.  This often affects the root screen when you create chats and navigate back without dismissing the keyboard.

I hope this was helpful.

File: src/MessageContainer.js [hunk 9]
```diff
@@ -95,34 +51,36 @@ export default class MessageContainer extends React.Component {
...
-      previousMessage: message.previousMessage,
-      nextMessage: message.nextMessage,
-      position: message.user._id === this.props.user._id ? 'right' : 'left',
+      ...restProps,
+      key: item._id,
+      currentMessage: item,
+      previousMessage,
+      nextMessage,
+      position: item.user._id === this.props.user._id ? 'right' : 'left',
+      hash: sum(
```
Comment (filbertteo) [inline src/MessageContainer.js]: What is the purpose of `hash`? As far as I'm aware, the `Message` component does not expect a `hash` prop.

**Comment by PeterKottas** [comment]
Hi gents, wow there's a lot of PRs flying around :) I was working on #768 and realized react-native-invertible-scroll-view doesn't work with web. Then I started looking around at ways of fixing it and found that apparently you removed the dependency from this branch altogether. Kind of good news for me I think :) But I wanted to check if my assumptions are correct and also the status of this. How confident do you feel this will end up in master? And I know I am a bit pushy (sorry) but do you have a clue about timelines for this? Thanks!

**Comment by PeterKottas** [comment]
Looking at the code here https://github.com/FaridSafi/react-native-gifted-chat/pull/705/commits/10b8e6a9ee183761277f72f85ba2e54560717677#diff-bd8395a0a94eca8b61a34ec66f8a605cR112 I have a bad feeling it might not work anyways. To me, the obvious solution would be to reverse messages order and put a flex-filler on top of the list. However, I am afraid that native implementation might not work with that. Any thoughts?

**Comment by PeterKottas** [comment]
That actually happens on current master as well. At least on android.

**Comment by PeterKottas** [comment]
Tried again and it actually works ok. The app probably wasn't rebuild properly. It does that on the expo example app though. Not sure what release that points to though.

**Comment by xcarpentier** [comment]
@PeterKottas yes, it's difficult to know which version is executed on expo, I think it will be great to add version / branch visible somewhere (header?)

BTW @Shelf-A @PeterKottas Do you think it's a bocker?

**Comment by PeterKottas** [comment]
Maybe we could do a commit hash as part of CI and make a link to it on github? We had something like that on a solution long time ago and it proved very popular with testers <-> developers relation.

**Comment by PeterKottas** [comment]
Also I am all for merging this asap cause this is a BIG piece of work and I am afraid of merge-hell if we wait too long.

**Comment by theplatapi** [comment]
Can this please be merged? I have a chat app out but it slows down after time due to the old list memory issues. Just using the flatlist will be a huge improvement.

**Comment by brunocascio** [comment]
@xcarpentier 
I have access to npm. Let me know when it is ready, and I'll deploy it!

**Comment by xcarpentier** [comment]
@brunocascio I have access too, but I think it will be great to make a v0.4.0 first, no?
With current master.

**Comment by brunocascio** [comment]
@xcarpentier How does it looks? :) https://github.com/FaridSafi/react-native-gifted-chat/releases/tag/v0.4.0

**Comment by xcarpentier** [comment]
@brunocascio OK for me

**Comment by brunocascio** [comment]
@xcarpentier I've just published v0.4.1 version. Go ahead :)

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
@Shelf-A Thanks for your feedback, I changed something:
1. Removed props on Flatlist
1. Added a condition on setState in InputToolbar.js

Can you have a look?

**Comment by xcarpentier** [comment]
@PeterKottas I just added the version in header of expo-example. ;)

**Comment by xcarpentier** [comment]
@theplatapi I want too but please test it here and let me know if you find any issues: https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

**Comment by Shelf-A** [comment]
It starts off perfectly then when the list grows/after sending a couple of messages the input/keyboard freezes as you type and also when you clear text.
screenshot_20180316-151624
Platform: android

**Comment by xcarpentier** [comment]
@Shelf-A So, we need to stop rendering of bunch list using FlatList props optimization somehow.

cc https://github.com/facebook/react-native/issues/13413

**Comment by xcarpentier** [comment]
Perf optimization in last version:
- removeClippedSubviews on android only
- Message extends PureComponent

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by xcarpentier** [comment]
@Shelf-A can you test the last one please? 🙄

**Comment by xcarpentier** [comment]
:shipit: This branch has been deployed to:
  https://expo.io/@giftedchat/react-native-gifted-chat-expo-modified-for-flatlist

  Download the [Expo](https://expo.io/) app and scan this QR code to get started!

  QR Code

**Comment by Shelf-A** [comment]
@xcarpentier the issue is fixed, works smooth

**Comment by xcarpentier** [comment]
Question:

Is that better to remove PureComponent to use shouldComponentUpdate instead because with a small scope conditional update by component (ie. regarding the component and child needs...)

Because PureComponent make a big comparison and it's not worthing for all. Should we consider using more shouldComponentUpdate or keep as is?

**Comment by Shelf-A** [comment]
I think we should consider using more shouldComponentUpdate, like for double ticks and time

**Comment by brunocascio** [comment]
React devs recommend `PureComponent` only on stateless components. Keep it in mind we can choose between `PureComponent and `Component` depending on how to perform it in your app.
`PureComponent` is just a `Component` that implements `shouldComponentUpdate` with shallow comparison.

**Comment by xcarpentier** [comment]
> `PureComponent` is just a `Component` that implements `shouldComponentUpdate` with shallow comparison.

Yes I know! But this is currently my question: why make a big shallow comparison (ie. with PureComponent)? Because in GC we passed all the props from the root to all leaf.

So I think it will be better for us to use `shouldComponentUpdate` to compare only props we need by components and children of this component or change our mind totally by not passing all props to all components.

- https://github.com/FaridSafi/react-native-gifted-chat/blob/10f0a42d5a3b73b4c84806b03d842acf909981fd/src/GiftedChat.js#L314
- https://github.com/FaridSafi/react-native-gifted-chat/blob/10f0a42d5a3b73b4c84806b03d842acf909981fd/src/GiftedChat.js#L436
- https://github.com/FaridSafi/react-native-gifted-chat/blob/10f0a42d5a3b73b4c84806b03d842acf909981fd/src/MessageContainer.js#L120
- https://github.com/FaridSafi/react-native-gifted-chat/blob/10f0a42d5a3b73b4c84806b03d842acf909981fd/src/InputToolbar.js#L68
- etc

So the *shallow comparison* is pretty huge, this is not just from a few props but all props.

What do you think?

(I think you should keep this way to not change to much but instead use more `shouldComponentUpdate` than `PureComponent` for perf reason)

**Comment by brunocascio** [comment]
Yes, I agree. I think that performance optimization could take place when those troubles appear. I tested it and it works, so we can release a new version in order to get some user feedbacks as well :)

**Comment by theplatapi** [comment]
Tested this branch out with my current app. Rendering a user's message after it's sent takes a little longer, but the app no longer slows down as the list grows. I'll use what's here for now since users are really complaining about the slowdowns!

Also, the load previous button doesn't show the animation when pressed, probably because it was moved to a pure component. Other than that a great improvement!

**Comment by brunocascio** [comment]
@xcarpentier The `example`  folder is already updated or what should I use for testing?

**Comment by helielson** [comment]
@xcarpentier any updates or ETA here? I'm still using the `modified-for-flatlist` branch 😆

File: src/MessageContainer.js [hunk 10]
```diff
@@ -131,36 +84,32 @@ export default class MessageContainer extends React.Component {
...
-          renderRow={this.renderRow}
-          renderHeader={this.props.inverted ? this.renderFooter : this.renderLoadEarlier}
-          renderFooter={this.props.inverted ? this.renderLoadEarlier : this.renderFooter}
-          renderScrollComponent={this.renderScrollComponent}
+          data={this.props.messages}
+          style={styles.listStyle}
+          contentContainerStyle={styles.contentContainerStyle}
+          renderItem={this.renderRow}
+          renderHeader={this.renderFooter}
+          renderFooter={this.renderLoadEarlier}
```
Comment (brunolemos) [inline src/MessageContainer.js]: Why is `renderHeader=renderFooter` and `renderFooter=renderLoadEarlier`? 🤪

**Comment by howg0924** [comment]
@brunolemos as for the TextInput does not move up problem, I fixed it in my own fork, maybe you could take it as a reference: https://github.com/howg0924/react-native-gifted-chat/commit/06e78bee77624d2a8755a3b072bed56208667fb9

I solved it by calling keyboard listeners directly in GiftedChat.js. I don't understand why these listeners were passed down to the FlatList in MessageContainer.js.

The problem only happens on iOS on my app.

**Comment by brunolemos** [comment]
@howg0924 unfortunately cherry-picking this commit did not solve the issue... maybe because your branch is outdated?

**Comment by howg0924** [comment]
@brunolemos: My fork is not outdated on this branch. I also fixed some other issues and did a little enhancement on my fork. I guess maybe these modification also helped to fixed the issue. Or maybe the root cause of your problem is different from mine.

**Comment by brunolemos** [comment]
@howg0924 yeah I merged master on it, maybe it was that

**Comment by b4stien** [comment]
Hello everyone, I am not sure what's blocking this PR, can we help in any way?

We're using this project at my company and would be willing to help land this (which seems to be fixing a rendering issue with messages not appearing until scroll).

**Comment by xcarpentier** [comment]
@b4stien some performance issue. Need to add some `suc` and maybe change some pure component

**Comment by sibelius** [comment]
@FaridSafi can we merge this, and track some regression in new issues?

**Comment by xcarpentier** [comment]
@sibelius I will merge it tomorrow.

**Comment by helielson** [comment]
Nice @xcarpentier! Can't wait! Let me know if you need any help

**Comment by jaburcodes** [comment]
awesome PR!! Thanks @xcarpentier pretty sure this will help a lot of devs!

**Comment by dead23angel** [comment]
bump version pls

File: src/InputToolbar.js [hunk 11]
```diff
@@ -33,15 +33,19 @@ export default class InputToolbar extends React.Component {
   }
 
   keyboardWillShow() {
-    this.setState({
-      position: 'relative',
-    });
+    if (this.state !== 'relative') {
```
Comment (varungupta85) [inline src/InputToolbar.js]: Should this be `this.state.position`? Same for the next if condition also.

**Comment by fbartho** [comment]
@varungupta85 that bug looks like it just got merged into master!

File: src/InputToolbar.js (hunk 11 above)
Comment (brunolemos) [reply src/InputToolbar.js]: This is probably what caused this bug: https://github.com/FaridSafi/react-native-gifted-chat/pull/705#pullrequestreview-120090347

**Comment by ChrisEdson** [comment]
Is it possible to get a release with the new flat list design?

**Comment by glebez** [comment]
Is it just me, or `rednerFooter` prop of gifted chat stopped properly working after this branch has been merged?

On first sight it seems to me, that `FlatList` does not expect props `renderHeader` and `renderFooter`, but instead `ListHeaderComponent` and `ListFooterComponent`, which makes them be ignored.

**Comment by xcarpentier** [comment]
@glebez Do you have any screenshot or code to share?
Or maybe think about create a PR with the fix...

**Comment by sibelius** [comment]
We should have an issue with all possible regressions of this PR

**Comment by glebez** [comment]
@xcarpentier, I'll try to put together a reduced test case first and afterwards a PR.

**Comment by glebez** [comment]
Here's a test case using current npm package, so before the FlatList was introduced and the footer is displayed properly: https://snack.expo.io/@glebkost/listview-example

If you change the package to use current master the footer is not displayed. (Unfortunately it's not possible in expo snack to set the package to github url, so you need to copy the snack to your local env to test that).

**Pull request merged**